### PR TITLE
Added B3Propagator for B3 Single Header

### LINKED ISSUE #500: Create B3 Propagator
**Is your feature request related to a problem?**
`W3C tracecontext` is not yet supported "everywere" plus B3 propagation is part of the [Spec Compliance Matrix](https://github.com/open-telemetry/opentelemetry-specification/blob/main/spec-compliance-matrix.md#context-propagation).

**Describe the solution you'd like**
Two Propagator implementations. One for B3 [Single Headers](https://github.com/openzipkin/b3-propagation#single-header), one for B3 [Multiple Headers](https://github.com/openzipkin/b3-propagation#multiple-headers).

### PR DESCRIPTION
**Pull request by kishannsangani**
Missed out implementing B3 Single Header in #680.

Closes #500.

### CODE REVIEW

**Comment by kishannsangani** [comment]
@Nevay Added B3Propagator Class to handle two configurations and stored debug flag in the returned context. However, since we have implemented a singleton class for all the propagators, it is not possible to test the B3Propagator Class since there is no mechanism to reset the singleton static instance.

PS: @tidal @bobstrecansky @brettmc

**Comment by Nevay** [comment]
We should allow users to choose between the two modes, either by having a singleton for each mode, or by exposing the constructor with an argument that allows choosing between the two modes.

[Configuration Interface](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/sdk-configuration.md#programmatic):
> The SDK MUST provide a programmatic interface for all configuration. This interface SHOULD be written in the language of the SDK itself. All other configuration mechanisms SHOULD be built on top of this interface.

**Comment by kishannsangani** [comment]
@Nevay so having a third B3Propagator class is not required since we have a singleton for each B3 Single and B3 Multi propagator?

**Comment by Nevay** [comment]
No, we still need the combined `::extract()` behavior - I meant replacing `B3Propagator::getInstance()` with a method for each configuration mode, something like `::getB3SingleHeaderInstance()` and `::getB3MultiHeaderInstance()`.

File: src/API/Trace/Propagation/B3Propagator.php [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+/**
+ * B3 is a propagator that supports the specification for the header
+ * "b3" used for trace context propagation across service boundaries.
+ * (https://github.com/openzipkin/b3-propagation)
+ */
+final class B3Propagator
+{
+    public static function getB3SingleHeaderInstance(): TextMapPropagatorInterface
+    {
+        return B3SinglePropagator::getInstance();
```
Comment (Nevay) [inline src/API/Trace/Propagation/B3Propagator.php:18]: `::getB3SingleHeaderInstance()` and `::getB3MultiHeaderInstance()` have to return a `B3Propagator` that uses both propagators for extracting, similar to `::getInstance()` before the last commit.
```php
private function __construct(TextMapPropagatorInterface $propagator)
{
	$this->propagator = propagator;
}
public static function getB3SingleHeaderInstance(): self
{
	static $instance;
	return $instance ??= new self(B3SinglePropagator::getInstance());
}
public static function getB3MultiHeaderInstance(): self
{
	static $instance;
	return $instance ??= new self(B3MultiPropagator::getInstance());
}
```

File: src/API/Trace/Propagation/B3SinglePropagator.php [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+/**
+ * B3 is a propagator that supports the specification for the B3 single
+ * header "b3" used for trace context propagation across service boundaries.
+ * (https://github.com/openzipkin/b3-propagation#single-header)
+ */
+final class B3SinglePropagator implements TextMapPropagatorInterface
+{
+    public const B3 = 'b3';
+
+    public static ContextKey $B3_DEBUG_FLAG_KEY;
```
Comment (Nevay) [inline src/API/Trace/Propagation/B3SinglePropagator.php:28]: This context key has to be shared between single header and multi header propagator.

File: src/API/Trace/Propagation/B3Propagator.php (hunk 1 above)
Comment (kishannsangani) [reply src/API/Trace/Propagation/B3Propagator.php]: Implemented.

File: src/API/Trace/Propagation/B3SinglePropagator.php (hunk 2 above)
Comment (kishannsangani) [reply src/API/Trace/Propagation/B3SinglePropagator.php]: Created new class `B3DebugFlagContextKey` as you suggested.

File: src/API/Trace/Propagation/B3Propagator.php [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+
+    /** {@inheritdoc} */
+    public function extract($carrier, PropagationGetterInterface $getter = null, Context $context = null): Context
+    {
+        $getter ??= ArrayAccessGetterSetter::getInstance();
+        $context ??= Context::getCurrent();
+
+        $b3Context = $getter->get($carrier, B3SinglePropagator::B3);
+
+        if ($b3Context) {
```
Comment (Nevay) [inline src/API/Trace/Propagation/B3Propagator.php:60]: I'm not sure whether this is sufficient; do we have to parse the multi-header if the `b3` single-header is present but invalid?

**Review by Nevay — COMMENTED** [review]
Should be moved out of the API into an extension package - see [Propagators Distribution](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/context/api-propagators.md#propagators-distribution).

File: src/API/Trace/Propagation/B3Propagator.php (hunk 3 above)
Comment (kishannsangani) [reply src/API/Trace/Propagation/B3Propagator.php:60]: The specs says `MUST attempt to extract` which is ambiguous about the the outcome.

**Comment by kishannsangani** [comment]
> Should be moved out of the API into an extension package - see [Propagators Distribution](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/context/api-propagators.md#propagators-distribution).

@Nevay do we have any extension packages in opentelemetry-php?

**Comment by brettmc** [comment]
> do we have any extension packages in opentelemetry-php?
discussed in SIG, please go ahead and create an extensions dir (this is what otel-java does)

**Comment by kishannsangani** [comment]
> > do we have any extension packages in opentelemetry-php?
> 
> discussed in SIG, please go ahead and create an extensions dir (this is what otel-java does)

@brettmc otel-java has nested directories under extension. Do we follow the same or a simple hierarchy? Also, they have api and sdk on the root folder whereas we have it under src. Should I place extension under the src?

**Comment by brettmc** [comment]
I think `src/Extension/Propagator/B3` or something similar...

**Comment by kishannsangani** [comment]
> @kishannsangani - do we feel this is ready now?

Yes, I just pushed the final changes for the clarification on [specs](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/context/api-propagators.md#b3-extract).

@Nevay @bobstrecansky @brettmc @tidal can you review it?

**Review by Nevay — APPROVED** [review]
Needs a `composer.json` in either `src/Extension/Propagator` or `src/Extension/Propagator/B3`.

**Comment by kishannsangani** [comment]
> Needs a `composer.json` in either `src/Extension/Propagator` or `src/Extension/Propagator/B3`.

@Nevay added `composer.json` in `src/Extension/Propagator/B3`. Can you help me verify the same?

File: src/Extension/Propagator/B3/composer.json [hunk 4]
```diff
@@ -0,0 +1,23 @@
+{
+    "name": "open-telemetry/extension/propagator/B3",
```
Comment (Nevay) [inline src/Extension/Propagator/B3/composer.json:2]:
```suggestion
    "name": "open-telemetry/extension-propagator-b3",
```

File: src/Extension/Propagator/B3/composer.json [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+    "description": "B3 propagator extension for OpenTelemetry PHP.",
+    "keywords": ["opentelemetry", "otel", "metrics", "tracing", "logging", "apm", "extension", "propagator", "b3"],
```
Comment (Nevay) [inline src/Extension/Propagator/B3/composer.json:4]:
```suggestion
    "keywords": ["opentelemetry", "otel", "tracing", "apm", "extension", "propagator", "b3"],
```

**Review by brettmc — APPROVED** [review]

**Pull request merged**